src/Actions/Commission/CreateCommissionCreditNotes.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace FluxErp\Actions\Commission;
+
+use FluxErp\Actions\FluxAction;
+use FluxErp\Actions\Order\CreateOrder;
+use FluxErp\Actions\OrderPosition\CreateOrderPosition;
+use FluxErp\Enums\OrderTypeEnum;
+use FluxErp\Models\Commission;
+use FluxErp\Models\Order;
+use FluxErp\Models\OrderPosition;
+use FluxErp\Models\OrderType;
+use FluxErp\Models\User;
+use FluxErp\Models\VatRate;
+use FluxErp\Rulesets\Commission\CreateCommissionCreditNotesRuleset;
+use FluxErp\Support\Collection\OrderCollection;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+class CreateCommissionCreditNotes extends FluxAction
+{
+    protected function boot(array $data): void
+    {
+        parent::boot($data);
+
+        $this->rules = resolve_static(CreateCommissionCreditNotesRuleset::class, 'getRules');
+    }
+
+    public static function models(): array
+    {
+        return [Commission::class];
+    }
+
+    public function performAction(): OrderCollection
+    {
+        $agents = User::query()
+            ->withWhereHas('contact', fn (Builder|BelongsTo $query) => $query->with('invoiceAddress.country'))
+            ->whereHas('commissions', fn (Builder|HasMany $query) => $query->whereIntegerInRaw('id', $this->data))
+            ->get(['id', 'name', 'email', 'contact_id']);
+        $orderIds = [];
+
+        foreach ($agents as $agent) {
+            $clientIds = resolve_static(OrderPosition::class, 'query')
+                ->whereHas('commission', fn (Builder $query) => $query->whereIntegerInRaw('id', $this->data)
+                    ->where('user_id', $agent->id)
+                )
+                ->distinct('client_id')
+                ->get('client_id')
+                ->pluck('client_id');
+
+            foreach ($clientIds as $clientId) {
+                $commissions = $agent->commissions()
+                    ->whereIntegerInRaw('id', $this->data)
+                    ->whereDoesntHave('creditNoteOrderPosition')
+                    ->withWhereHas(
+                        'orderPosition',
+                        fn (Builder|BelongsTo $query) => $query->where('client_id', $clientId)
```

Comment:
why is this Builder|BelongsTo as parameter type?

Reply:
see comment above